File: public/collections.pl
#!/usr/bin/perl

use lib "..";
use strict;

use Flavors::Data::Collection;
use Flavors::Data::Playlist;
use Flavors::Data::Song;
use Flavors::Data::Tag;
use Flavors::Data::Util;
use Flavors::HTML;
use Flavors::Util;
use List::Util qw(shuffle);
use POSIX qw(floor);

my $dbh = Flavors::Data::Util::DBH();

my $cgi = CGI->new;
print $cgi->header();
my $fdat = Flavors::Util::Fdat($cgi);
Flavors::HTML::Header($dbh, {
    TITLE => "Collections",
    BUTTONS => Flavors::HTML::ExportControl() . qq{
        <button type="button" class="btn btn-xs btn-info" id="add-collection">
            <span class="fas fa-plus"></span>
            New
        </button>
    },
    JS => ['collections.js', 'song_attributes.js', 'filters.js', 'playlists.js'],
});

my $results = Flavors::Data::Util::TrySQL($dbh, {
    SUB => 'Flavors::Data::Collection::List',
    ARGS => {
        FILTER => $fdat->{FILTER},
        ORDERBY => $fdat->{ORDERBY},
        UPDATEPLAYLIST => 1,
    },
});
my $sqlerror = $results->{ERROR} || "";
my @collections = @{ $results->{RESULTS} };

my %songs;
my @songs = Flavors::Data::Song::List($dbh);
foreach my $song (@songs) {
    $songs{$song->{ID}} = $song;
}

print Flavors::HTML::FilterControl($dbh, {
    FILTER => $fdat->{FILTER},
    ERROR => $sqlerror,
    HASCOUNT => 1,
    HASRANDOM => 1,
});

print qq{ <div class="post-nav"> };

print "<div class=\"collections clearfix\">";

my %colors;
foreach my $color (Flavors::Data::Tag::ColorList($dbh)) {
    $colors{$color->{NAME}} = $color;
}

foreach my $collection (@collections) {
    printf(qq{
            <div 
                data-id="%s"
                data-original-title="%s"
                data-tag-list="%s"
                data-name="%s"
                data-artist="%s"
                data-artist-list="%s"
                data-starred="%s"
                data-song-count="%s"
                data-total="%s"
                class="collection clearfix"
            >
        },
        $collection->{ID},
        $collection->{NAME},
        Flavors::Util::EscapeHTMLAttribute(join(" ", @{ $collection->{TAGS} })),
        lc($collection->{NAME}),
        lc($collection->{ARTIST}),
        lc($collection->{ARTISTLIST}),
        $collection->{ISSTARRED} ? 1 : 0,
        $collection->{SONGCOUNT},
        $collection->{TOTAL},
    );

    my @files = Flavors::Data::Collection::CoverArtFiles($collection->{ID});
    if (@files && !$collection->{ISMIX}) {
        @files = shuffle(@files);
        printf(qq{
                <div class="cover-art%s">
                    %s
                </div>
            },
            @files > 1 ? " multiple" : "",
            join("", map { sprintf("<img src='%s' />", $_) } @files[0 .. (scalar(@files) < 4 ? scalar(@files) - 1 : 4)]));
    }
    else {
        my $color = $colors{@{ $collection->{COLORS} }[0]};
        my $acronym = $collection->{NAME};
        $acronym =~ s/[^\w\s]//g;
        $acronym = join("", map { substr($_, 0, 1) } split(' ', $acronym));
        my $acronymsize = "solo";
        if (length($acronym) >= 8) {
            $acronymsize = "xsmall";
        }
        elsif (length($acronym) >= 5) {
            $acronymsize = "small";
        }
        elsif (length($acronym) == 4) {
            $acronymsize = "medium";
        }
        elsif (length($acronym) == 3) {
            $acronymsize = "large";
        }
        elsif (length($acronym) == 2) {
            $acronymsize = "xlarge";
        }
        printf(qq{
                <div class="cover-art missing" style="%s%s">
                    <div class="acronym-%s">%s</div>
                </div>
            },
            $color ? ("background-color: #" . $color->{HEX} . ";") : "",
            $color->{WHITETEXT} ? " color: white; font-weight: bold;" : "",
            $acronymsize,
            $acronym,
        );
    }
    printf(qq{
            <div class="accepting-drop hide">
                <i class="fas fa-cloud-upload-alt"></i>
                <br /><br />
                Drop new cover art
            </div>
            <div class="name">%s</div>
            <div class="artist">%s</div>
        },
        $collection->{NAME},
        $collection->{ARTIST}
    );

    my $exporttext = "";
    if ($collection->{EXPORTCOUNT} == 0) {
        $exporttext = "Never exported";
    }
    else {
        $exporttext = "Exported ";
        if ($collection->{EXPORTCOUNT} == 1) {
            $exporttext .= "once, on ";
        }
        else {
            if ($collection->{EXPORTCOUNT} == 2) {
                $exporttext .= "twice";
            }
            else {
                $exporttext .= $collection->{EXPORTCOUNT} . " times";
            }
            $exporttext .= "<br>Last exported ";
        }
        $exporttext .= " " . Flavors::Util::TrimDate($collection->{LASTEXPORT});
    }

    printf(qq{
            <div class="details-background">
            <div class="details">
                <div>Acquired %s</div>
                %s<br><br>
                <div class="ratings">
                    <div>
                        <div class="rating">%s</div>
                        <div class="rating">%s</div>
                        <div class="rating">%s</div>
                    </div>
                    <div>
                        <div class="rating">%s</div>
                        <div class="rating">%s</div>
                        <div class="rating">%s</div>
                    </div>
                    <div>
                        <div class="rating">%s</div>
                        <div class="rating">%s</div>
                        <div class="rating">%s</div>
                    </div>
                    <div>%s</div>
                </div>
                <div class="tags">%s</div>
            </div>
            </div>
            <ul class="cover-art-thumbnails clearfix hide">%s</ul>
        },
        Flavors::Util::TrimDate($collection->{CREATED}),
        $exporttext,
        Flavors::HTML::Rating($collection->{MINRATING}, 'star'),
        Flavors::HTML::Rating($collection->{MINENERGY}, 'fire'),
        Flavors::HTML::Rating($collection->{MINMOOD}, 'heart'),
        Flavors::HTML::Rating($collection->{AVGRATING}, 'star'),
        Flavors::HTML::Rating($collection->{AVGENERGY}, 'fire'),
        Flavors::HTML::Rating($collection->{AVGMOOD}, 'heart'),
        Flavors::HTML::Rating($collection->{MAXRATING}, 'star'),
        Flavors::HTML::Rating($collection->{MAXENERGY}, 'fire'),
        Flavors::HTML::Rating($collection->{MAXMOOD}, 'heart'),
        $collection->{COMPLETION} == 1 ? "&nbsp;" : sprintf("(%s%% complete)", floor($collection->{COMPLETION} * 100)),
        join("", map { "<div>$_</div>" } @{ $collection->{TAGS} }[0..2]),
        join("", map { sprintf("<li><img src='%s' /><div class='trash'><i class='fas fa-trash'></i></div></li>", $_) } @files),
    );

    print "</div>";
}
print "</div></div>";

# Modal for new collection
print q{
    <div id="new-collection" class="modal">
        <div class="modal-dialog modal-lg">
            <div class="modal-content">
                <div class="modal-header">
                    <button type="button" class="close" data-dismiss="modal">
                        <span aria-hidden="true">&times;</span><span class="sr-only">Close</span>
                    </button>
                    <h4>
                        <input type='text' name='name' placeholder='collection' />
                        <label>
                            <input type='checkbox' name='ismix' />
                            is mix
                        </label>
                    </h4>
                </div>
                <div class="modal-body">
                    <table>
                        <tr class='song hide'>
                            <td class='ordinal'>0</td>
                            <td><input type='text' name='name' placeholder='song' /></td>
                            <td><input type='text' name='artist' placeholder='artist' /></td>
                            <td>
                                <input type='text' name='minutes' placeholder='0' />
                                :
                                <input type='text' name='seconds' placeholder='00' />
                            </td>
                            <td><span class='fas fa-trash'></span></td>
                        </tr>
                    </table>
                    <div id="add-song">
                        <button class='btn btn-default btn-large'>
                            <span class='fas fa-plus'></span>
                        </button>
                        <input type='text' value='1' />
                    </div>
                </div>
                <div class="modal-footer">
                    <button id='cancel-add-collection' class='btn btn-default btn-large'>
                        cancel
                    </button>
                    <button id='save-collection' class='btn btn-primary btn-large'>
                        save
                    </button>
                </div>
            </div>
        </div>
    </div>
};

print Flavors::HTML::SongsModal();

print Flavors::HTML::Footer();

File: public/javascript/collections.js
function addSong(focus) {
    var $modal = jQuery("#new-collection");
    var lastArtist = $modal.find(".song [name='artist']:last").val();
    var $newSong = $modal.find(".song.hide").clone().removeClass("hide");
    $modal.find("#add-song").before($newSong);
    if (lastArtist) {
        $newSong.find("[name='artist']").val(lastArtist);
    }
    var $rows = $modal.find(".song:visible");
    _.each($rows, function(row, r) {
        _.each($(row).find("input"), function(input, i) {
            if (i < 2) {
                // First two columns: name and artist
                $(input).attr("tabindex", $rows.length * i + r + 1);
            } else {
                // Time columns
                $(input).attr("tabindex", $rows.length * 2 + r + i - 1);
            }
        });
    });
    if (focus) {
        $newSong.find("input:first").focus();
    }
    numberSongs();
}

function dropCollection(id, name) {
    var $well = jQuery(".controls .well");
    if (!$well.find("li[data-id='" + id + "']").length)  {
        var $ul = $well.find("ul");
        $well.find(".subtle").addClass("hide");
        var li = "<li data-id=\"" + id + "\">";
        li += name;
        li += "</li>";
        $ul.append(li);
        jQuery(".collections .collection[data-id='" + id + "']").addClass("selected");
    }
}

function numberSongs() {
    jQuery("#new-collection .ordinal:visible").each(function(index) {
        jQuery(this).html(index + 1 + '.');
    });
}

function filterCollections(force) {
    var query = jQuery("#filter").val().toLowerCase();
    var selector = ".collections .collection";
    var queryTokens = _.without(query.split(/\s+/), "");
    if (!queryTokens.length) {
        jQuery(selector).show();
    }
    else {
        jQuery(selector).show();
        _.each(queryTokens, function(queryToken) {
            jQuery(selector + ":visible").each(function() {
                var $collection = jQuery(this);
                var haystack = _.map(["name", "artist", "artist-list", "tag-list"], function(a) {
                    return $collection.attr("data-" + a);
                }).join(" ").toLowerCase();
                if (haystack.indexOf(queryToken) === -1) {
                    $collection.hide();
                }
            });
        });
    }

    // Account for stars
    if (!jQuery("#simple-filter .far.fa-star").length) {
           jQuery(selector + "[data-starred!='1']").hide();
    }

    // Randomize order if requested
    if (!jQuery("#simple-filter .fa-random").hasClass("text-muted")) {
        var $container = $(".collections"),
            collections = $container.find(".collection").detach();
        collections = _.sortBy(collections, function() { return Math.random(); });
        $container.append(collections);
    }

    return jQuery(".collections .collection:visible").length;
}

jQuery(document).ready(function() {
    jQuery(".collection").click(function() {
        var $collection = jQuery(this),
            id = $collection.data("id");
        showSongModal({
            TITLE: $collection.find(".name").text(),
            SUB: 'Flavors::Data::Collection::TrackList', 
            COLLECTIONIDS: id,
        }, function() {
            var $modal = $("#song-list");
            $modal.data("id", id);
            if ($collection.find(".cover-art.multiple").length) {
                $modal.find(".modal-body").append($collection.find(".cover-art-thumbnails").clone().removeClass("hide"));
            }

            var $backdrop = $(".modal-backdrop.in"),
                $image = $backdrop.clone(),
                $covers = $collection.find("img");
            $image.css("background-color", "transparent")
                  .css("background-size", "cover")
                  .css("background-repeat", "no-repeat")
                  .css("background-position", "center");
            if ($covers.length) {
                var src = $covers[parseInt(Math.random() * $covers.length)].src;
                $image.css("background-image", "url('" + src + "')")
            }
            $backdrop.after($image);

            $modal.one("hide.bs.modal", function() {
                $image.remove();
            });
        });
    });

    // Column names hint for filter
    jQuery(".hint").tooltip({
        html: true,
        placement: "right"
    });

    initSimpleFilter(filterCollections);

    // Controls: Add collection
    jQuery("#add-collection").click(function() {
        var $modal = jQuery("#new-collection");
        $modal.modal();
        $modal.find("input:first").focus();
        if (!$modal.find(".song:visible").length) {
            addSong(false);
        }
    });

    jQuery("#add-song input").focus(function() {
        jQuery(this).select();
    });

    jQuery("#new-collection #add-song button").click(function() {
        var count = jQuery("#add-song input").val();
        if (!parseInt(count)) {
            count = 1;
        }
        for (var i = 0; i < count; i++) {
            addSong(true);
        }
    });

    jQuery("#new-collection").on("click", ".fa-trash", function() {
        jQuery(this).closest(".song").remove();
        numberSongs();
    });

    jQuery("#cancel-add-collection").click(function() {
        var $modal = jQuery("#new-collection");
        $modal.find(".song:visible").remove();
        $modal.find("input[type='text']").val("");
        $modal.find("input[type='checkbox']").attr("checked", false);
        $modal.modal('hide');
    });

    jQuery("#save-collection").click(function() {
        var data = {};
        var $modal = jQuery("#new-collection");
        var $header = $modal.find(".modal-header");
        var $name = $header.find("[name='name']");
        if (!$name.val()) {
            alert("Missing name");
            $name.focus();
        }

        var args = {
            NAME: $name.val(),
            ISMIX: $header.find("input[type='checkbox']").attr("checked"),
            SONGS: [],
        };
        var attributes = ['name', 'artist', 'minutes', 'seconds'];
        $modal.find(".song:visible").each(function() {
            var values = {};
            for (var i = 0; i < attributes.length; i++) {
                var $obj = jQuery(this).find("[name='" + attributes[i] + "']");
                if (!$obj.val()) {
                    alert("Missing data");
                    $obj.focus();
                }
                values[attributes[i].toUpperCase()] = $obj.val();
            }
            if (values.SECONDS < 10) {
                values.SECONDS = '0' + values.SECONDS;
            }
            values.TIME = values.MINUTES + ':' + values.SECONDS;
            args.SONGS.push(values);
        });
        jQuery(this).attr("disabled", true);
        CallRemote({
            SUB: 'Flavors::Data::Collection::Add',
            ARGS: args,
            FINISH: function() {
                location.reload();
            }
        });
    });

    // Export single collection, from modal
    jQuery("#song-list .export-dropdown a").click(function(event) {
        var $collection = jQuery(".collection[data-id='" + jQuery(this).closest("#song-list").data("id") + "']");
        var $details = $collection.find(".details");
        ExportPlaylist({
            CONFIG: jQuery(this).data("name"),
            COLLECTIONIDS: [$collection.data("id")],
            FILENAME: $collection.find(".artist").text() + " - " + $collection.find(".name").text(),
        });
    });

    // Page-level export set of collections
    jQuery("nav .export-dropdown a").click(function() {
        var $button = jQuery(this);
        var collectionids = [];
        jQuery(".collection:visible").each(function() {
            collectionids.push(jQuery(this).data("id"));
        });
        if (!collectionids.length) {
            alert("No collections selected");
            return;
        }
        ExportPlaylist({
            CONFIG: jQuery(this).data("name"),
            COLLECTIONIDS: collectionids,
        });
    });

    // Upload cover art
    var div = document.createElement('div');
    if ((('draggable' in div) || ('ondragstart' in div && 'ondrop' in div)) && 'FormData' in window && 'FileReader' in window) {
        var $targets = jQuery(".collection");
        $targets.on('drag dragstart dragend dragover dragenter dragleave drop', function(e) {
            e.preventDefault();
            e.stopPropagation();
        });
        var current = undefined;
        $targets.on('dragover dragenter', function(e) {
            current = jQuery(e.currentTarget).data("id");
            jQuery('.accepting-drop').addClass('hide');
            jQuery(this).find('.accepting-drop').removeClass("hide");
        });
        $targets.on('dragleave dragend drop', function(e) {
            if (jQuery(e.currentTarget).data("id") !== current) {
                jQuery(this).find('.accepting-drop').addClass("hide");
                current = undefined;
            }
        });
        var formats = [];
        $targets.on('drop', function(e) {
            var $collection = jQuery(this),
                id = $collection.data("id"),
                data = new FormData();

            $collection.find('.accepting-drop').addClass('hide');

            if (!confirm("Upload new art for " + $collection.data("originalTitle") + "?")) {
                return;
            }
            if (e.originalEvent.dataTransfer.files.length !== 1) {
                alert("Please drag a single file.");
                return;
            }
            var file = e.originalEvent.dataTransfer.files[0];
            if (file.type !== "image/png" && file.type !== "image/jpeg") {
                alert("File must be either a PNG or JPEG.");
                return;
            }
            data.append('file', file);
            data.append('ext', file.type.replace(/.*\//, '').replace(/e/, ''));
            data.append('id', id);
            data.append('sub', 'Flavors::Data::Collection::UpdateCoverArt');

            CallRemote({
                ARGS: data,
                FINISH: function(data) {
                    if (data.FILENAME) {
                        var $art = $collection.find(".cover-art"),
                            $img = jQuery("<img />");
                        $img.attr("src", data.FILENAME + "?" + (new Date()).getTime());
                        if ($art.hasClass("missing")) {
                            $art.removeClass("missing");
                            $art.css("background-color", "");
                            $art.html("");
                        } else {
                            $art.addClass("multiple");
                        }
                        $art.append($img);
                        var $li = jQuery("<li></li>");
                        $li.append($img.clone());
                        $collection.find(".cover-art-thumbnails").append($li);
                    }
                },
                UPLOAD: true,
            });
        });
    }

    // Remove cover art
    $(document).on('click', '.cover-art-thumbnails .trash', function() {
        if (confirm('Remove image?')) {
            var $li = $(this).closest("li"),
                src = $li.find("img").attr("src"),
                id = $("#song-list").data("id");
            CallRemote({
                SUB: 'Flavors::Data::Collection::RemoveCoverArt',
                ARGS: {
                    id: id,
                    filename: src,
                },
                FINISH: function(data) {
                    if (data.FILENAME === src) {
                        $li.remove();
                        $collection = $(".collection[data-id='" + id + "']");
                        $collection.find(".cover-art img[src='" + src + "']").remove();
                        $collection.find(".cover-art-thumbnails img[src='" + src + "']").closest("li").remove();
                    } else {
                        alert("Unable to remove image");
                    }
                }
            });
        }
    });
});
